parachain/pallets/ethereum-beacon-client/src/lib.rs

```diff
@@ -595,6 +658,71 @@ pub mod pallet {
 			Ok(())
 		}
 
+		fn ancestry_proof(
+			block_root_proof: BoundedVec<H256, T::MaxProofBranchSize>,
+			block_slot: u64,
+			beacon_block_root: H256,
+			finalized_header_root: H256,
+		) -> DispatchResult {
+			// If the block root proof is empty, we know that we expect this header to be a
+			// finalized header. We need to check that the header hash matches the finalized header
+			// root at the expected slot.
+			if block_root_proof.len() == 0 {
+				let stored_finalized_header = <FinalizedBeaconHeaders<T>>::get(beacon_block_root);
+				if stored_finalized_header.is_none() {
+					log::error!(
+						target: "ethereum-beacon-client",
+						"💫 Finalized block root {} slot {} for ancestry proof (for a finalized header) not found.", beacon_block_root, block_slot
+					);
+					return Err(Error::<T>::ExpectedFinalizedHeaderNotStored.into())
+				}
+
+				let header = stored_finalized_header.unwrap();
+				if header.slot != block_slot {
+					log::error!(
+						target: "ethereum-beacon-client",
+						"💫 Finalized block root {} slot {} does not match expected slot {}.", beacon_block_root, block_slot, header.slot
+					);
+					return Err(Error::<T>::UnexpectedHeaderSlotPosition.into())
+				}
+
+				return Ok(())
+			}
+
+			let finalized_block_root_hash =
+				<FinalizedBeaconHeadersBlockRoot<T>>::get(finalized_header_root);
+
+			if finalized_block_root_hash.is_zero() {
+				log::error!(
+					target: "ethereum-beacon-client",
+					"💫 Finalized block root {} slot {} for ancestry proof not found.", beacon_block_root, block_slot
+				);
+				return Err(Error::<T>::ExpectedFinalizedHeaderNotStored.into())
+			}
+
+			let max_slots_per_historical_root = T::MaxSlotsPerHistoricalRoot::get() as u64;
```

Comment:
If we are type casting like this, shouldn't T::MaxSlotsPerHistoricalRoot be `u64` in the first place?